File: ExpressionParser/Core/Tokenizer.cs
using Expresser.Core.Exceptions;
using Expresser.Core.Tokens;

namespace Expresser.Core;

internal static class Tokenizer
{
    private static readonly HashSet<char> _operatorsSymbols = new() { '<', '>', '=', '+', '-', '*', '/' };
    private static readonly HashSet<string> _literalOperators = new() { "not", "or", "and" };
    private static readonly HashSet<string> _functions = new() { "mulTen" };

    public static Span<Token> Tokenize(string stringExpression)
    {
        var tokens = new List<Token>();
        var pos = 0;
        for (; pos < stringExpression.Length; pos++)
        {
            var ch = stringExpression[pos];
            var charType = GetCharType(ch);
            var token = charType switch
            {
                CharType.Digit => GetNumberConstantToken(stringExpression, ref pos),
                CharType.StringConstantSymbol => GetStringConstantToken(stringExpression, ref pos),
                CharType.OperatorSymbol => GetOperatorToken(stringExpression, ref pos, tokens),
                CharType.Letter => GetTokenFromString(stringExpression, ref pos),
                CharType.OpeningBracket => new OpeningBracket(),
                CharType.ClosingBracket => new ClosingBracket(),
                CharType.WhiteSpace => null,
                _ => throw new ParsingException($"Unknown symbol {ch}. Position: {pos}")
            };

            if (token != null)
            {
                tokens.Add(token);
            }
        }

        return tokens.ToArray().AsSpan();
    }

    private static NumberConstant GetNumberConstantToken(string expression, ref int pos)
    {
        var startPosition = pos;
        var strNumber = new ValueStringBuilder(16);

        for (; pos < expression.Length; pos++)
        {
            char ch = expression[pos];
            if (char.IsDigit(ch) || ch == '.')
                strNumber.Append(ch);
            else
            {
                pos--;
                break;
            }
        }

        if (double.TryParse(strNumber.AsSpan(), out var value))
        {
            return new NumberConstant(value);
        }
        else
        {
            throw new ParsingException($"Can`t parse {strNumber.ToString()} to number constant. Position: {startPosition}");
        }
    }

    private static StringConstant GetStringConstantToken(string stringExpression, ref int pos)
    {
        var startPosition = pos;
        var value = new ValueStringBuilder(16);

        pos++;//skip first "
        for (; pos < stringExpression.Length; pos++)
        {
            char ch = stringExpression[pos];
            if (ch == '\\')
            {
                if (pos + 1 < stringExpression.Length && stringExpression[pos + 1] == '\\')
                {
                    value.Append('\\');
                    pos++;
                    continue;
                }
                if (pos + 1 < stringExpression.Length && stringExpression[pos + 1] == '"')
                {
                    value.Append('"');
                    pos++;
                    continue;
                }
                
                throw new ParsingException($"Unexpected escape sequence. Expected \\\" or \\\\ . Position: {pos}");
            }
            if (ch == '"')
            {
                return new StringConstant(value.ToString());
            }
            else
            {
                value.Append(ch);
            }
        }

        throw new ParsingException($"String constant has no end. Symbol \" expected in the end of string constant. Position: {startPosition}");
    }

    private static Token GetTokenFromString(string expression, ref int pos)
    {
        var stringBuilder = new ValueStringBuilder(16);
        var dotPosition = -1;

        for (; pos < expression.Length; pos++)
        {
            char ch = expression[pos];
            if (char.IsLetter(ch))
            {
                stringBuilder.Append(ch);
            }
            else if (ch == '.')
            {
                if (dotPosition == -1)
                {
                    dotPosition = pos;
                    stringBuilder.Append(ch);
                }
                else
                {
                    throw new ParsingException($"Variable property must be simple type. Unexpected symbol '.' in Position {pos}");
                }
            }
            else
            {
                pos--;
                break;
            }
        }

        var value = stringBuilder.ToString();

        return value switch
        {
            "true" => new BooleanConstant(true),
            "false" => new BooleanConstant(false),
            "not" => new Not(),
            "or" => new Or(),
            "and" => new And(),
            "mulTen" => new MulTen(),
            _ when dotPosition != -1 => new ParameterPropertyInvocation(value.Substring(0, dotPosition), value.Substring(dotPosition + 1)),
            _ => new Parameter(value)
        };
    }

    private static Token GetOperatorToken(string stringExpression, ref int pos, IEnumerable<Token> previousTokens)
    {
        var op = "";
        if (pos < stringExpression.Length - 1
            && _operatorsSymbols.Contains(stringExpression[pos + 1]))
        {
            op = $"{stringExpression[pos]}{stringExpression[pos + 1]}";
            pos++;
        }
        else
        {
            op = stringExpression[pos].ToString();
        }

        return op switch
        {
            "-" when previousTokens.LastOrDefault() is not { } prevToken 
                || (prevToken is BinaryOperator or OpeningBracket or Function) => new Negate(),
            "=" => new Equal(),
            "<>" => new NotEqual(),
            ">" => new GreaterThan(),
            "<" => new LessThan(),
            ">=" => new GreaterThanOrEqual(),
            "<=" => new LessThanOrEqual(),
            "+" => new Add(),
            "-" => new Subtract(),
            "*" => new Multiply(),
            "/" => new Divide(),
            _ => throw new ParsingException($"Undefined operator {op}. Position: {pos}")
        };
    }

    private static CharType GetCharType(char c)
    {
        if (char.IsDigit(c))
            return CharType.Digit;

        if (char.IsLetter(c))
            return CharType.Letter;

        if (char.IsWhiteSpace(c))
            return CharType.WhiteSpace;

        if (c == '"')
        {
            return CharType.StringConstantSymbol;
        }

        if (c == '(')
        {
            return CharType.OpeningBracket;
        }

        if (c == ')')
        {
            return CharType.ClosingBracket;
        }

        if (_operatorsSymbols.Contains(c))
            return CharType.OperatorSymbol;

        return CharType.Unknown;
    }

    internal enum CharType
    {
        Unknown,
        Digit,
        Letter,
        OperatorSymbol,
        StringConstantSymbol,
        OpeningBracket,
        ClosingBracket,
        WhiteSpace
    }
}

File: Expresser.FSharp/Tokenizer.fs
namespace Expresser.Fsharp

module Tokenizer =

    type Token =
        | OpeningBracket
        | ClosingBracket
        | Parameter of string
        | ParameterPropertyAccess of ParameterName: string * PropertyName: string
        | NumberConstant of double
        | StringConstant of string
        | BooleanConstant of bool
        | Negate
        | Not
        | Add
        | Subtract
        | Multiply
        | Divide
        | Or
        | And
        | GreaterThan
        | LessThan
        | GreaterThanOrEqual
        | LessThanOrEqual
        | Equal
        | NotEqual
        | Sin
        | Cos

    let (|BinaryOperatorToken|_|) token =
        match token with
        | Add | Subtract | Multiply | Divide | Or | And | GreaterThan | LessThan | GreaterThanOrEqual | LessThanOrEqual | Equal | NotEqual -> Some token
        | _ -> None

    let (|UnaryOperatorToken|_|) token =
        match token with
        | Negate | Not -> Some token
        | _ -> None

    let (|FunctionToken|_|) token =
        match token with
        | Sin | Cos -> Some token
        | _ -> None

    let (|ConstantToken|_|) token =
        match token with
        | NumberConstant _ | StringConstant _ | BooleanConstant _ -> Some token
        | _ -> None

    let Tokenize stringExpression = 

        let GetStringConstantToken (stringExpression: char list) tokens =
            let rec GetStringConstant stringExpression value = 
                match stringExpression with
                | '/'::'/'::tail -> GetStringConstant tail (value @ ['/'])
                | '/'::'"'::tail -> GetStringConstant tail (value @ ['"'])
                | '/'::unknown::_ -> failwith $"Unexpected escape sequence \\{unknown}. Expected \\\" or \\\\"
                | '"'::_-> value
                | char::tail -> GetStringConstant tail (value @ [char])
                | [] -> failwith "String constant has no end. Symbol \" expected in the end of string constant"
            let constant = GetStringConstant stringExpression [] |> List.map string |> List.reduce (+)
            let token = constant |> StringConstant
            (stringExpression[constant.Length+1..], tokens @ [token])

        let (|Double|_|) str =
            match System.Double.TryParse (str:string) with
            | true, double -> Some double
            | _ -> None

        let (|Digit|_|) char =
            match System.Char.IsDigit char with
            | true -> Some char
            | _ -> None

        let (|Letter|_|) char =
            match System.Char.IsLetter char with
            | true -> Some char
            | _ -> None

        let (|OperatorChar|_|) char =
            match List.contains char ['<'; '>'; '='; '+'; '-'; '*'; '/'] with
            | true -> Some char
            | _ -> None

        let GetNumberConstantToken stringExpression tokens =
            let rec GetNumberConstant stringExpression value = 
                match stringExpression with
                | Digit digit::tail -> GetNumberConstant tail (value @ [digit])
                | '.'::tail -> GetNumberConstant tail (value @ ['.'])
                | _ -> value
            let constant = GetNumberConstant stringExpression [] |> List.map string |> List.reduce (+)
            let number = 
                match constant with
                | Double double -> double
                | _ -> failwith $"Can`t parse {constant} to number constant."
            let token = number |> NumberConstant
            (stringExpression[constant.Length..], tokens @ [token])

        let (|Letters|_|) string =
            match String.forall (fun x -> System.Char.IsDigit x) string with
            | true -> Some string
            | _ -> None

        let (|ParameterPropertyAccess|_|) (str: string) =
            let splittedString = str.Split [|'.'|] |> Seq.toList
            match splittedString with
            | [ Letters parameter; Letters property ] -> Some (parameter, property)
            | _ -> None

        let GetTokenFromString stringExpression tokens =
            let rec GetTokenFromStringStep stringExpression value =
                match stringExpression with
                | Letter letter :: tail -> GetTokenFromStringStep tail (value @ [letter])
                | '.'::tail -> GetTokenFromStringStep tail (value @ ['.'])
                | _ -> value
            let string = GetTokenFromStringStep stringExpression [] |> List.map string |> List.reduce (+)
            let token =
                match string with
                | "true" -> BooleanConstant true
                | "false" -> BooleanConstant false
                | "not" -> Not
                | "Or" -> Or
                | "And" -> And
                | "sin" -> Sin
                | "cos" -> Cos
                | Letters parameter -> Parameter parameter
                | ParameterPropertyAccess parameterPropertyAccess -> ParameterPropertyAccess parameterPropertyAccess
                | _ -> failwith $"Can`t find token '{string}'"
            (stringExpression[string.Length..], tokens @ [token])

        let GetOperatorToken (stringExpression:char list) (tokens: Token list) =
            let (token, tokenLength) = 
                match stringExpression with
                | '-'::_ when (tokens.Length > 0 && 
                    match List.last tokens with
                    | BinaryOperatorToken _ | FunctionToken _ | OpeningBracket -> true
                    | _ -> false) -> (Negate, 1)
                | '<'::'>'::_ -> (NotEqual, 2)
                | '>'::'='::_ -> (GreaterThanOrEqual, 2)
                | '<'::'='::_ -> (LessThanOrEqual, 2)
                | '='::_ -> (Equal, 1)
                | '>'::_ -> (GreaterThan, 1)
                | '<'::_ -> (LessThan, 1)
                | '+'::_ -> (Add, 1)
                | '-'::_ -> (Subtract, 1)
                | '*'::_ -> (Multiply, 1)
                | '/'::_ -> (Divide, 1)
                | _ -> failwith "Unexpected operator token"
            (stringExpression[tokenLength..], tokens @ [token])

        let rec TokenizeStep (stringExpression: char list) (tokens: Token list) =
            match stringExpression with
            | '"'::other -> GetStringConstantToken other tokens ||> TokenizeStep
            | '('::other -> TokenizeStep other (tokens @ [OpeningBracket])
            | ')'::other -> TokenizeStep other (tokens @ [ClosingBracket])
            | ' '::other -> TokenizeStep other tokens
            | Digit _::_ -> GetNumberConstantToken stringExpression tokens ||> TokenizeStep
            | Letter _::_-> GetTokenFromString stringExpression tokens ||> TokenizeStep
            | OperatorChar _::_-> GetOperatorToken stringExpression tokens ||> TokenizeStep
            | [] -> tokens
            | unknownChar::_ -> invalidOp $"Unknown symbol {unknownChar}"
        
        TokenizeStep stringExpression []